charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,189 @@
+        self.charm = charm
+        self.relation_endpoint = relation_endpoint
+        self.status = self._get_status()
+
+        if charm.unit.is_leader():
+            self.framework.observe(
+                charm.on[relation_endpoint].relation_created, self._check_provider
+            )
+            self.framework.observe(
+                charm.on[relation_endpoint].relation_changed, self._check_provider
+            )
+            self.framework.observe(charm.on.leader_elected, self._check_provider)
+
+    def _get_status(self):
+        if not self.charm.unit.is_leader():
+            return None
+        try:
+            ingress = get_interface(self.charm, self.relation_endpoint)
+            if not ingress:
+                return BlockedStatus(f"Missing relation: {self.relation_endpoint}")
+        except NoCompatibleVersions:
+            return BlockedStatus(f"Relation version not compatible: {self.relation_endpoint}")
+        except NoVersionsListed:
+            return WaitingStatus(f"Waiting on relation: {self.relation_endpoint}")
+        else:
+            return None
+
+    @property
+    def is_available(self):
+        return self.charm.unit.is_leader() and self.status is None
+
+    @property
+    def is_ready(self):
+        return self.is_available and self.url
+
+    def _check_provider(self, event):
+        if self.is_ready:
+            self.on.ready.emit()
+        elif self.is_available:
+            self.on.available.emit()
+        elif isinstance(self.status, BlockedStatus):
+            self.on.failed.emit()
+
+    def request(
+        self,
+        *,
+        port: int,
+        service: str = None,
+        prefix: str = None,
+        rewrite: str = None,
+        namespace: str = None,
+        per_unit_routes: bool = False,
+    ):
+        """Request ingress to a service.
+
+        Args:
+            service: the name of the target K8s service to route to; defaults to the
+                charm's automatically created service (i.e., the application name)
+            port: the port of the service (required)
+            prefix: the path used to match this service for requests to the gateway;
+                must not conflict with other services; defaults to f"/{service}/"
+            rewrite: the path on the target service to map the request to; defaults
+                to "/"
+            namespace: the namespace the service is in; default to the current model
+            per_unit_routes: whether or not to create URLs which map to specific units;
+                the URLs will have their own prefix of f"{prefix}-unit-{unit_num}" (with
+                tailing slashes handled appropriately)
+        """
+        if not self.charm.unit.is_leader():
+            raise RequestFailed(
+                WaitingStatus(f"Only leader can request ingress: {self.relation_endpoint}")
+            )
+        if self.status is not None:
+            raise RequestFailed(self.status)
```

Comment:
I actually would expect the requirer charm to specify these parameters in the constructor of `IngressRequirer`. That tends to make the usage of the library more resilient, as you do not need to hook it specifically anywhere in your charm, just instantiate `IngressRequirer` in your charm's constructor.

Reply:
Apologies; I don't understand this comment. Creating the instance of the `IngressRequirer` in the charm's `__init__` is exactly the pattern being used.  I also don't know what "parameters" you're referring to here, as this is just error checking and propagating an apparently ignored bad status as an exception with the same info.

Reply:
So, what the code seems to imply is that you will have to invoke the `request` method somewhere in a hook. I am saying that the same parameters you pass to `request` you could rather be passing to the `IngressRequirer` constructor (https://github.com/canonical/istio-operators/pull/43/files#diff-e846165da73fe4674f0e581c3df778471bec5f1c5caab26f06437772fd4eeda8R49), and do not expect the charm to invoke any method to "trigger" the requesting of an ingress mapping: that would happen automatically when the relation between requirer charm and the ingress charm is established.

Reply:
That would restrict a give charm to requesting ingress to only a single service per relation to the ingress provider. That seems like an unnecessary arbitrary restriction, given that there's no reason a particular operator charm can't deploy multiple services or even define services dynamically based on things like other relations or config. (Granted, the way the underlying schema is currently defined also imposes this limitation, but now that you've called that out, I think I should fix the new schema PR rather than changing how this part works).

Reply:
Actually, I think your good intentions will result in a confusing UX for charm authors. In the case of wanting to expose independently different parts of a charm’s workload API, it makes for a better DevX (and also admin UX), to have separate relations using the `ingress` relation interface. That is, in the prometheus-k8s charm, we would declare two relations with the `ingress` relation interface, named, for example, `ingress-remote-write` and `ingress-web-ui`.

The alternative is… what? Having the charm author put boolean switches in their charm’s config to give the Juju admin the option of deciding which APis to expose? That would be super un-Juju :-)

The model I propose also means that your code can remain super simple :-)

Reply:
Supporting the request / response pattern does not preclude the charm author from defining multiple relation endpoints, and we did just that in the CK charms for the `loadbalancer` interface because it made sense in that context.  But there could well be other contexts where it would make more sense for a charm to expose a group of services with the same ingress characteristics.

Also, this fundamentally *is* a request / response pattern, because the properties of the ingress request must be sent over once the version negotiation is successful, before the response can be returned. Hiding that in the constructor doesn't change that fact and just obscures the actual flow of logic for minimal or no gain in usability. (Whether the params are passed in to the constructor or a method, they still have to be passed in. And doing so in a handler gives the charm author the chance to set an appropriate status, should they want, which they can't do under your proposal.)

All that said, this is not a hill I'm willing to die on, because I'm still just trying to get this thing working and useful at this point. :sweat_smile:

Reply:
Added optional request init param support so it can be used with this pattern, or with an explicit call to `request`.